### ISSUE TITLE
MultiOptimizer is not sufficient to freeze layer/separate training of different components

### ISSUE THREAD

**Issue by ChikaYan** [google/flax#1764]
I'd like to use Adam to train a model with two separate sets (A and B) of layers in an alternative fashion. I.e., for batch 1 only the layers in set A are trained and set B is frozen, for batch 2 set B is trained and set A is frozen. (Batches are carefully designed so batch 1,3,5... contain specific data that should only contribute to the training of set A)

The issue [here](https://github.com/google/flax/issues/825) suggests that this can be achieved with [MultiOptimizer](https://flax.readthedocs.io/en/latest/flax.optim.html#multioptimizer) by using one optimizer for each of the set and alternatively passing `learning_rate = lr` or `learning_rate = 0` to them. But after experimenting with this approach, I realized the training is not separated at all and layers from both sets seem to be trained with all the batches given. I suspect this is because of the momentum feature of Adam optimizer, even though the  `learning_rate = 0` is set, the layers are not completely frozen.

May I ask if there is another way to ensure the layers are properly frozen? Many thanks.

**Comment by andsteing** [comment]
We recommend using `optax` for optimization. The `flax.optim` package is effectively deprecated.

In `optax` you could achieve what you want via `optax.masked()` or `optax.multi_transform()`. See the function pydoc for examples.

**Comment by ChikaYan** [comment]
Thx a lot!
